### PR TITLE
[dbus] fix and enhance energy scan api

### PR DESCRIPTION
**Pull request by Irving-cl**
Fixes a bug in energy scan api and enhance according to @simonlingoogle 's comment.

### CODE REVIEW

File: src/utils/thread_helper.cpp [hunk 1]
```diff
@@ -136,7 +136,7 @@ void ThreadHelper::EnergyScan(uint32_t aScanDuration, EnergyScanHandler aHandler
     otError  error             = OT_ERROR_NONE;
     uint32_t preferredChannels = otPlatRadioGetPreferredChannelMask(mInstance);
 
-    VerifyOrExit(aHandler != nullptr, error = OT_ERROR_BUSY);
+    VerifyOrExit(aHandler != nullptr, error = OT_ERROR_INVALID_ARGS);
     VerifyOrExit(aScanDuration < UINT16_MAX, error = OT_ERROR_INVALID_ARGS);
     mEnergyScanHandler = aHandler;
```
Comment (simonlingoogle) [inline src/utils/thread_helper.cpp:141]: If `mEnergyScanHandler` was not `nullptr`, would it be a OT_ERROR_BUSY?

File: src/utils/thread_helper.cpp [hunk 2]
```diff
@@ -147,7 +147,7 @@ void ThreadHelper::EnergyScan(uint32_t aScanDuration, EnergyScanHandler aHandler
 exit:
     if (error != OT_ERROR_NONE)
     {
-        if (aHandler)
+        if (mEnergyScanHandler)
```
Comment (simonlingoogle) [inline src/utils/thread_helper.cpp:150]: When `error != OT_ERROR_NONE`, `mEnergyScanHandler` is likely the callback set by previous `EnergyScan`. Why are we calling the callback of previous `EnergyScan` ?

File: src/utils/thread_helper.cpp [hunk 3]
```diff
@@ -199,10 +199,10 @@ void ThreadHelper::EnergyScanCallback(otEnergyScanResult *aResult)
...
     if (aResult == nullptr)
     {
-        if (mEnergyScanHandler != nullptr)
-        {
-            mEnergyScanHandler(OT_ERROR_NONE, mEnergyScanResults);
-        }
+        assert(mEnergyScanHandler != nullptr);
+
+        mEnergyScanHandler(OT_ERROR_NONE, mEnergyScanResults);
+        mEnergyScanHandler = nullptr;
```
Comment (simonlingoogle) [inline src/utils/thread_helper.cpp:205]: We should be extremely careful when calling callbacks. The callback may be reset during the handler. 

```suggestion
        auto callback = mEnergyScanHandler;
        mEnergyScanHandler = nullptr;
        callback(OT_ERROR_NONE, mEnergyScanResults);
```

File: src/utils/thread_helper.cpp [hunk 4]
```diff
@@ -199,10 +199,10 @@ void ThreadHelper::EnergyScanCallback(otEnergyScanResult *aResult)
 {
     if (aResult == nullptr)
     {
-        if (mEnergyScanHandler != nullptr)
-        {
-            mEnergyScanHandler(OT_ERROR_NONE, mEnergyScanResults);
-        }
+        assert(mEnergyScanHandler != nullptr);
+
+        mEnergyScanHandler(OT_ERROR_NONE, mEnergyScanResults);
```
Comment (simonlingoogle) [inline src/utils/thread_helper.cpp:204]: should also clear mEnergyScanResults after callback?

File: src/utils/thread_helper.cpp [hunk 5]
```diff
@@ -136,7 +136,7 @@ void ThreadHelper::EnergyScan(uint32_t aScanDuration, EnergyScanHandler aHandler
     otError  error             = OT_ERROR_NONE;
     uint32_t preferredChannels = otPlatRadioGetPreferredChannelMask(mInstance);
 
-    VerifyOrExit(aHandler != nullptr, error = OT_ERROR_BUSY);
+    VerifyOrExit(aHandler != nullptr, error = OT_ERROR_INVALID_ARGS);
```
Comment (jinran-google) [inline src/utils/thread_helper.cpp:139]: We'd better add test cases covering all these error cases, like null handler, scanning again before the previous one ends, starting scan in the callback, etc.

File: src/utils/thread_helper.cpp [hunk 6]
```diff
@@ -147,7 +147,7 @@ void ThreadHelper::EnergyScan(uint32_t aScanDuration, EnergyScanHandler aHandler
 exit:
     if (error != OT_ERROR_NONE)
```
Comment (simonlingoogle) [inline src/utils/thread_helper.cpp:148]: Usually, we install callbacks only when the error is `OT_ERROR_NONE` , and only then the callbacks will be invoked later. 
Why are we using a different strategy here, where callbacks seem to be called no matter it's successful or not?
